File: pkg/store/user_update_store.go
package store

import (
	"context"
	"fmt"
	"time"

	"gorm.io/gorm"

	"github.com/PineappleBond/xyncra-server/pkg/store/model"
)

// UserUpdateStore provides data access operations for the UserUpdate model.
type UserUpdateStore struct {
	db *gorm.DB
}

// Create inserts a batch of user update records using CreateInBatches (batch
// size 100) for efficient bulk insertion.
func (us *UserUpdateStore) Create(ctx context.Context, updates []model.UserUpdate) error {
	if len(updates) == 0 {
		return nil
	}
	if err := us.db.WithContext(ctx).CreateInBatches(updates, 100).Error; err != nil {
		return classifyError(err)
	}
	return nil
}

// ListByUser returns user updates for the given userID with Seq greater than
// afterSeq, ordered by Seq ascending, limited to at most limit rows.
func (us *UserUpdateStore) ListByUser(ctx context.Context, userID string, afterSeq uint32, limit int) ([]*model.UserUpdate, error) {
	if limit <= 0 || limit > 1000 {
		limit = 100
	}

	var updates []*model.UserUpdate
	err := us.db.WithContext(ctx).
		Where("user_id = ? AND seq > ?", userID, afterSeq).
		Order("seq ASC").
		Limit(limit).
		Find(&updates).Error
	if err != nil {
		return nil, classifyError(fmt.Errorf("store: list user updates: %w", err))
	}
	return updates, nil
}

// ListByUserRange returns user updates for the given userID with Seq in the
// range (afterSeq, maxSeq] (exclusive start, inclusive end), ordered by Seq
// ascending.
func (us *UserUpdateStore) ListByUserRange(ctx context.Context, userID string, afterSeq, maxSeq uint32) ([]*model.UserUpdate, error) {
	if maxSeq <= afterSeq {
		return nil, nil
	}
	var updates []*model.UserUpdate
	err := us.db.WithContext(ctx).
		Where("user_id = ? AND seq > ? AND seq <= ?", userID, afterSeq, maxSeq).
		Order("seq ASC").
		Find(&updates).Error
	if err != nil {
		return nil, classifyError(fmt.Errorf("store: list user updates by range: %w", err))
	}
	return updates, nil
}

// GetLatestSeq returns the highest Seq value for the given user. Returns 0 if
// the user has no update records.
func (us *UserUpdateStore) GetLatestSeq(ctx context.Context, userID string) (uint32, error) {
	var seq uint32
	err := us.db.WithContext(ctx).
		Model(&model.UserUpdate{}).
		Where("user_id = ?", userID).
		Select("COALESCE(MAX(seq), 0)").
		Scan(&seq).Error
	if err != nil {
		return 0, classifyError(fmt.Errorf("store: get latest seq: %w", err))
	}
	return seq, nil
}

// DefaultCleanupRetention is the default retention period for user updates.
const DefaultCleanupRetention = 30 * 24 * time.Hour // 30 days

// CleanupExpiredBefore hard-deletes all user updates with CreatedAt strictly
// before the given time. Returns the number of deleted rows.
func (us *UserUpdateStore) CleanupExpiredBefore(ctx context.Context, before time.Time) (int64, error) {
	result := us.db.WithContext(ctx).
		Unscoped().
		Where("created_at < ?", before).
		Delete(&model.UserUpdate{})
	if result.Error != nil {
		return 0, classifyError(fmt.Errorf("store: cleanup expired user updates: %w", result.Error))
	}
	return result.RowsAffected, nil
}

// CleanupExpired hard-deletes all user updates older than DefaultCleanupRetention
// (30 days). Convenience wrapper around CleanupExpiredBefore.
func (us *UserUpdateStore) CleanupExpired(ctx context.Context) (int64, error) {
	return us.CleanupExpiredBefore(ctx, time.Now().Add(-DefaultCleanupRetention))
}

File: demo/web/packages/xyncra-client-core/src/db/user-update-store.ts
/**
 * UserUpdateStore — data access operations for the UserUpdate model.
 *
 * Mirrors Go UserUpdateStore (pkg/store/user_update_store.go).
 *
 * UserUpdate does not use soft delete — expired records are hard-deleted
 * during cleanup (D-016).
 */

import type { XyncraDatabase } from './index';
import type { UserUpdate } from './models';

/** Default retention period for user updates: 30 days in milliseconds. */
const DEFAULT_CLEANUP_RETENTION_MS = 30 * 24 * 60 * 60 * 1000;

/**
 * UserUpdateStore provides data access operations for the UserUpdate model.
 */
export class UserUpdateStore {
  constructor(private readonly db: XyncraDatabase) {}

  /**
   * Inserts a batch of user update records for efficient bulk insertion.
   */
  async create(updates: UserUpdate[]): Promise<void> {
    if (updates.length === 0) return;
    await this.db.userUpdates.bulkAdd(updates);
  }

  /**
   * Returns user updates with seq greater than afterSeq,
   * ordered by seq ascending, limited to at most limit rows.
   *
   * Note: Database is already scoped by userID+deviceID, so no user_id filtering needed.
   */
  async listByUser(
    _userID: string,
    afterSeq: number,
    limit: number,
  ): Promise<UserUpdate[]> {
    if (limit <= 0 || limit > 1000) limit = 100;

    // Database is single-user scoped; just filter by seq.
    const updates = await this.db.userUpdates
      .where('seq')
      .above(afterSeq)
      .toArray();

    // Sort by seq ascending.
    updates.sort((a, b) => a.seq - b.seq);

    return updates.slice(0, limit);
  }

  /**
   * Returns user updates with seq in the range (afterSeq, maxSeq]
   * (exclusive start, inclusive end), ordered by seq ascending.
   *
   * Note: Database is already scoped by userID+deviceID, so no user_id filtering needed.
   */
  async listByUserRange(
    _userID: string,
    afterSeq: number,
    maxSeq: number,
  ): Promise<UserUpdate[]> {
    if (maxSeq <= afterSeq) return [];

    // Database is single-user scoped; just filter by seq range.
    const updates = await this.db.userUpdates
      .where('seq')
      .between(afterSeq + 1, maxSeq, true, true)
      .toArray();

    // Sort by seq ascending.
    updates.sort((a, b) => a.seq - b.seq);

    return updates;
  }

  /**
   * Returns the highest seq value. Returns 0 if no update records exist.
   *
   * Note: Database is already scoped by userID+deviceID, so no user_id filtering needed.
   */
  async getLatestSeq(_userID: string): Promise<number> {
    const updates = await this.db.userUpdates.toArray();

    if (updates.length === 0) return 0;
    return Math.max(...updates.map((u) => u.seq));
  }

  /**
   * Hard-deletes all user updates with created_at strictly before the given time.
   * Returns the number of deleted rows.
   */
  async cleanupExpiredBefore(before: Date): Promise<number> {
    const beforeMs = before.getTime();
    const count = await this.db.userUpdates
      .where('created_at')
      .below(before)
      .delete();
    return count;
  }

  /**
   * Hard-deletes all user updates older than DEFAULT_CLEANUP_RETENTION_MS (30 days).
   * Convenience wrapper around cleanupExpiredBefore.
   */
  async cleanupExpired(): Promise<number> {
    const before = new Date(Date.now() - DEFAULT_CLEANUP_RETENTION_MS);
    return this.cleanupExpiredBefore(before);
  }
}
